File: backend/app/schemas/knowledge_base.py
from datetime import datetime
from typing import Optional

from pydantic import BaseModel, Field


class KnowledgeArticleBase(BaseModel):
    title: str = Field(..., min_length=1, max_length=300)
    content: str = Field(..., min_length=1)
    category: str = Field(..., pattern="^(troubleshooting|how-to|faq|technical)$")
    language: str = Field(default="uk", pattern="^(uk|en)$")
    tags: Optional[list[str]] = None
    station_models: Optional[list[str]] = None
    error_codes: Optional[list[str]] = None


class KnowledgeArticleCreate(KnowledgeArticleBase):
    is_published: bool = False


class KnowledgeArticleUpdate(BaseModel):
    title: Optional[str] = Field(None, min_length=1, max_length=300)
    content: Optional[str] = Field(None, min_length=1)
    category: Optional[str] = Field(None, pattern="^(troubleshooting|how-to|faq|technical)$")
    language: Optional[str] = Field(None, pattern="^(uk|en)$")
    tags: Optional[list[str]] = None
    station_models: Optional[list[str]] = None
    error_codes: Optional[list[str]] = None
    is_published: Optional[bool] = None


class AuthorShort(BaseModel):
    id: int
    first_name: str
    last_name: str

    class Config:
        from_attributes = True


class KnowledgeArticleResponse(BaseModel):
    id: int
    title: str
    content: str
    category: str
    language: str
    tags: Optional[list[str]]
    station_models: Optional[list[str]]
    error_codes: Optional[list[str]]
    author_id: int
    author: AuthorShort
    is_published: bool
    view_count: int
    helpful_count: int
    created_at: datetime
    updated_at: Optional[datetime]

    class Config:
        from_attributes = True


class KnowledgeArticleListResponse(BaseModel):
    id: int
    title: str
    category: str
    language: str
    tags: Optional[list[str]]
    author: AuthorShort
    is_published: bool
    view_count: int
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class KnowledgeSearchRequest(BaseModel):
    query: str = Field(..., min_length=1)
    category: Optional[str] = None
    tags: Optional[list[str]] = None
    language: Optional[str] = Field(None, pattern="^(uk|en)$")
    limit: int = Field(default=50, ge=1, le=100)


class KnowledgeSearchResult(BaseModel):
    article_id: int
    title: str
    category: str
    content_preview: str
    score: float


class KnowledgeSearchResponse(BaseModel):
    results: list[KnowledgeSearchResult]
    query: str

File: frontend/src/api/knowledgeBase.ts
import client from './client'
import { PaginatedResponse } from './tickets'

export interface KnowledgeArticle {
  id: number
  title: string
  content?: string
  content_html?: string | null
  category: string
  language: string
  tags?: string[]
  error_codes?: string[]
  station_models?: string[]
  is_published: boolean
  is_featured?: boolean
  view_count: number
  helpful_count?: number
  not_helpful_count?: number
  author_id?: number
  author?: {
    id: number
    first_name: string
    last_name: string
  }
  last_editor_id?: number | null
  published_at?: string | null
  created_at: string
  updated_at: string
}

export interface ArticleListParams {
  page?: number
  per_page?: number
  search?: string
  category?: string
  language?: string
  is_published?: boolean
  tags?: string[]
}

export interface CreateArticleData {
  title: string
  content: string
  category: string
  language?: string
  tags?: string[]
  error_codes?: string[]
  station_models?: string[]
  is_published?: boolean
  is_featured?: boolean
}

export interface SearchResult {
  article: KnowledgeArticle
  score: number
  snippet: string
}

interface KnowledgeSearchResponse {
  results: Array<{
    article_id: number
    title: string
    category: string
    content_preview: string
    score: number
  }>
  query: string
}

export const knowledgeBaseApi = {
  list: async (params?: ArticleListParams): Promise<PaginatedResponse<KnowledgeArticle>> => {
    const response = await client.get<PaginatedResponse<KnowledgeArticle>>('/knowledge', { params })
    return response.data
  },

  get: async (id: number): Promise<KnowledgeArticle> => {
    const response = await client.get<KnowledgeArticle>(`/knowledge/${id}`)
    return response.data
  },

  create: async (data: CreateArticleData): Promise<KnowledgeArticle> => {
    const response = await client.post<KnowledgeArticle>('/knowledge', data)
    return response.data
  },

  update: async (id: number, data: Partial<CreateArticleData>): Promise<KnowledgeArticle> => {
    const response = await client.put<KnowledgeArticle>(`/knowledge/${id}`, data)
    return response.data
  },

  delete: async (id: number): Promise<void> => {
    await client.delete(`/knowledge/${id}`)
  },

  search: async (query: string, limit?: number, category?: string, tags?: string[], language?: string): Promise<SearchResult[]> => {
    const response = await client.post<KnowledgeSearchResponse>('/knowledge/search', {
      query,
      limit: limit || 50,
      category,
      tags,
      language,
    })

    return response.data.results.map((result) => ({
      article: {
        id: result.article_id,
        title: result.title,
        category: result.category,
        language: language || 'uk',
        content: result.content_preview,
        tags,
        is_published: true,
        view_count: 0,
        created_at: '',
        updated_at: '',
      },
      score: result.score,
      snippet: result.content_preview,
    }))
  },

  markHelpful: async (id: number, helpful: boolean): Promise<void> => {
    await client.post(`/knowledge/${id}/helpful`, { helpful })
  },

  getCategories: async (): Promise<string[]> => {
    const response = await client.get<string[]>('/knowledge/categories')
    return response.data
  },
}
